Add replace function to overwrite datasets

stuartmcalpine opened the pull request:
The way dataset overwriting was was a bit cumbersome, as whilst you were overwriting the data in the root_dir, a new entry was made in the database with a new version. It's a bit cleaner if the user really wants to overwrite a dataset that they can keep the same name/version/etc combination. Also in the past, the user needed to provide the `relative_path` of the dataset they are overwriting, which they would have to know before hand.

Now there is a `replace()` function for datasets, which mirrors exactly the inputs to the `register()` function.  They no longer need to pass the `relative_path` at any point.

Now when registering, if a dataset with the same `name`, `version_string`, `version_suffix`, `owner` and `owner_type` already exists it will raise an error. If they want to overwrite the dataset, they call the `replace()` function as they would the `register()` function. Naturally if the dataset they are trying to replace doesn't exist, or it is not overwritable, it raises an error.

Datasets that are overrwitten are tagged with the date and UID of who performed the operation, and the `dataset_id` that has replaced it.

### To do

- Update docs 
- ~Have `replace()` work when the `relative_path` was manually passed during `register()`~
- As before, any execution or dependencies links to the previous dataset will be lost

JoanneBogart reviewed (changes requested):
I like the idea of having a separate `replace` function.  I think the code would be easier to read and maintain if, instead of channeling everything into the function  `_register_or_replace`, you put the implementations in the separate `register` and `replace` functions.   Large sections of code which are common to both should be extracted to non-public functions.   So, for example, lines 148-208 (or perhaps 148-210)  could be in a function `_validate_register_inputs`.   The code from 212-265 is different for the two cases but then there is another large chunk (lines 267-376) which should be extracted to a utility function to be called by both `register` and `replace`

stuartmcalpine commented:
> I like the idea of having a separate `replace` function. I think the code would be easier to read and maintain if, instead of channeling everything into the function `_register_or_replace`, you put the implementations in the separate `register` and `replace` functions. Large sections of code which are common to both should be extracted to non-public functions. So, for example, lines 148-208 (or perhaps 148-210) could be in a function `_validate_register_inputs`. The code from 212-265 is different for the two cases but then there is another large chunk (lines 267-376) which could be extracted to a new utility function to be called by both `register` and `replace`

This was largely to prevent duplicating all the input options between a `register` function and a `replace` function. Maybe it's better to leave just `register` and have a `overwrite=True` argument in the function to mimic the replacing?

stuartmcalpine commented:
I have added:

- The `postgresql_nulls_not_distinct=True` for unique constraints, so we can have nullable columns in there
- Extracted some of the validation checks in `register()` to their own internal helper functions
- Added a `replace_iteration` column to datasets to track how many times a dataset has been overwritten, and added it to the unique constraints

stuartmcalpine commented:
Now the `replace` function is standalone. Both `register` and `replace` call some internal helper functions to validate their inputs and register the new row.

The tests weren't actually checking if they found a result, just checking the result assuming there was one, so I've added an initial check to see if it found them first.

stuartmcalpine commented:
Updated the first tutorial notebook to include replace, needs a check

JoanneBogart commented on src/dataregistry/registrar/dataset.py:
```diff
@@ -606,10 +708,10 @@ def _find_previous(self, name, version_string, version_suffix, owner, owner_type
             conn.commit()
 
-        # Pull out the single result
+        # There should only be a single result that is not overwritten
```
It's true that there *should* only be one, but it seems possible - albeit very unlikely - that if some earlier attempt to overwrite didn't complete properly or if the code was buggy that there could be more.  To be safe, I would add a clause to the query to order by `replace_iteration`, descending, and let the `replace_iteration`return value be a list.   Then `replace` will be able to find the correct next value for `replace_iteration` easily and, if there was some earlier issue with the value of `is_overwritten`, it can be silently fixed (or, perhaps better,  issue a warning message).   With the current code the replace will fail and the only way to fix it will be for the user to notify one of us to fix it manually.

stuartmcalpine replied:
`find_previous` now returns a dict, with the list entries ordered by `replace_iteration`.

Also, only `valid==1` datasets can now be replaced, i.e., valid but non-deleted, non-archived datasets.

Added a test to make sure deleted datasets cannot be replaced.

Added a test to make sure you cant replace is_overwritable=False

JoanneBogart commented on docs/source/tutorial_notebooks/getting_started_1_register_datasets.ipynb:
```diff
@@ -299,53 +295,54 @@
+    "Note that whilst the data in the shared space will be overwritten, the original entries in the data registry database are never lost (a new unique entry is created each time, and the 'old' entries will obtain `True` for their `is_overwritten` column and point to the dataset that overwrote them in the `replace_id` column). \n",
     "\n",
-    "For 2., to update a previous dataset and overwrite the existing data in the `root_dir`, we have to pass the `relative_path` of the existing dataset (see Section 6 for more details on the `relative_path`). For example"
+    "We can replace the first entry we made with different data as follow, note that `name`, `version`, `owner` and `owner_type` need to match our first entry to find the correct entry to replace."
```
Suggest this sentence be changed to

> We can replace the first entry we made with different data as shown in the next cell. Note that `name`, `version`, `owner` and `owner_type` need to match our first entry to find the correct entry to replace.

stuartmcalpine replied:
Changed

JoanneBogart commented on docs/source/tutorial_notebooks/getting_started_1_register_datasets.ipynb:
```diff
@@ -420,7 +417,7 @@
    "metadata": {},
    "source": [
-    "## 6) Special cases\n",
+    "## 7) Special cases\n",
```
I would modify the end of the first sentence under the heading "Registering external datasets" as follows:

> ..and to physically manage the data (either copy the cata to the central `root_dir` or verify that it already exists there).

JoanneBogart commented on src/dataregistry/registrar/dataset.py:
```diff
@@ -270,203 +412,163 @@ def _register_or_replace(
-                ds_creation_date,
-            ) = self._handle_data(
-                relative_path, old_location, owner, owner_type, verbose
+        # If `relative_path` not passed, automatically generate it
```
Shouldn't there also be a check when `relative_path` is passed in that it matches the expected value?

stuartmcalpine replied:
Ah this check shouldn't actually be needed at all, i've removed it. The relative path used is the same as the previous dataset (returned in found_previous), so it always gets overwritten in the code further down.

JoanneBogart replied:
Ok, that seems right.  But then it shouldn't even be an argument. Please take it out of the argument list.

JoanneBogart reviewed (changes requested):
This is close, but I think there are a couple things `replace` isn't checking that it should check.  I also have a quibble with the `_find_previous` behavior, and found a couple places where text could be clearer.

stuartmcalpine commented on src/dataregistry/registrar/dataset.py:
```diff
@@ -38,42 +38,25 @@ def __init__(self, db_connection, root_dir, owner, owner_type, execution_table):
         If something is invalid, an exception is raised.
 
+        Some properties may be overwritten given the value of other properties,
+        or automatically generated, hence why some properties are returned back
```
Ah this was left over from a previous version, updated with recommended.

JoanneBogart reviewed (changes requested):
I noted a couple specific things.  In addition, there are several aspects of the register and replace functions that may need some further work, not in all cases necessarily due to the changes coming from this PR, so I'll write about it in an issue.
See https://github.com/LSSTDESC/dataregistry/issues/144

stuartmcalpine commented on src/dataregistry/registrar/dataset.py:
```diff
@@ -666,24 +674,21 @@ def _find_previous(self, name, version_string, version_suffix, owner, owner_type
+        Returns results a dict stating what iteration those datasets are, the
+        relative path their data is located, and if they can be overwritten.
+
+        Only "valid", non-deleted, non-archived datasets are returned, i.e.,
```
Yes this was left over from before, changed

stuartmcalpine commented on src/dataregistry/registrar/dataset.py:
```diff
@@ -704,23 +710,28 @@ def _find_previous(self, name, version_string, version_suffix, owner, owner_type
+            for att in _return_atts:
+                data[att].append(getattr(r, att))
+
+        return data
```
Yes there was no real need.

I now return the `results.all()` which is the list of row objects, and then work with that in the `register()` and `replace()` functions.

JoanneBogart reviewed (approved):
There is one small thing to fix: get rid of unused, unneeded `relative_path` argument for replace. Other than that I think we're done for this round. It will be easier to see what's left to do with a fresh start from here.
